sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c

```diff
@@ -0,0 +1,304 @@
+{
+  UNUSED(context);
+  UNUSED(topicLen);
+
+  int i;
+  char* payloadptr;
+
+  printf("Message arrived\n");
+  printf("     topic: %s\n", topicName);
+  printf("   message: ");
+
+  payloadptr = message->payload;
+  for (i = 0; i < message->payloadlen; i++)
+  {
+    putchar(*payloadptr++);
+  }
+  putchar('\n');
+  MQTTClient_freeMessage(&message);
+  MQTTClient_free(topicName);
+  return 1;
+}
+
+static int connect()
+{
+  int rc;
+
+  MQTTClient_SSLOptions mqtt_ssl_options = MQTTClient_SSLOptions_initializer;
+  MQTTClient_connectOptions mqtt_connect_options = MQTTClient_connectOptions_initializer;
+
+  char username[128];
+  az_span username_span = AZ_SPAN_LITERAL_FROM_BUFFER(username);
+  if ((rc = az_iot_provisioning_client_user_name_get(
+           &provisioning_client, username_span, &username_span))
+      != AZ_OK)
+  {
+    printf("Failed to get MQTT username, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(username_span, '\0', &username_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT username, return code %d\n", rc);
+    return rc;
+  }
+
+  mqtt_connect_options.username = (char*)az_span_ptr(username_span);
+  mqtt_connect_options.password = NULL; // Using X509 Client Certificate authentication.
+
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to connect, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int subscribe()
+{
+  int rc;
+
+  char topic_filter[128];
+  az_span topic_filter_span = AZ_SPAN_LITERAL_FROM_BUFFER(topic_filter);
+  if ((rc = az_iot_provisioning_client_register_subscribe_topic_filter_get(
+           &provisioning_client, topic_filter_span, &topic_filter_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(topic_filter_span, '\0', &topic_filter_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_subscribe(mqtt_client, topic_filter, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int register_device()
+{
+  int rc;
+  MQTTClient_message pubmsg = MQTTClient_message_initializer;
+
+  char topic[128];
+  az_span topic_span = AZ_SPAN_LITERAL_FROM_BUFFER(topic);
+
+  if ((rc = az_iot_provisioning_client_register_publish_topic_get(
+           &provisioning_client, topic_span, &topic_span))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT PUB register topic, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = az_span_append_uint8(topic_span, '\0', &topic_span)) != AZ_OK)
+  {
+    printf("Failed to get MQTT PUB register topic, return code %d\n", rc);
```

Comment:
not exactly the error here

Reply:
:unamused: Not exactly the code I'd like someone else to write either.

Reply:
haha i hear ya

Reply:
I won't change before we decide on the public API.

Reply:
I'll follow-up with the change in a future PR.